crates/ruff_linter/src/rules/pyupgrade/rules/unpacked_list_comprehension.rs

```diff
@@ -0,0 +1,37 @@
+use ruff_diagnostics::Violation;
+use ruff_macros::{derive_message_formats, violation};
+
+/// ## Removed
+/// There's no [evidence](https://github.com/astral-sh/ruff/issues/12754) that generators are
+/// meaningfully faster than list comprehensions when combined with unpacking.
+///
+/// ## What it does
+/// Checks for list comprehensions that are immediately unpacked.
+///
+/// ## Why is this bad?
+/// There is no reason to use a list comprehension if the result is immediately
+/// unpacked. Instead, use a generator expression, which avoids allocating
+/// an intermediary list.
+///
+/// ## Example
+/// ```python
+/// a, b, c = [foo(x) for x in items]
+/// ```
+///
+/// Use instead:
+/// ```python
+/// a, b, c = (foo(x) for x in items)
+/// ```
+///
+/// ## References
+/// - [Python documentation: Generator expressions](https://docs.python.org/3/reference/expressions.html#generator-expressions)
+/// - [Python documentation: List comprehensions](https://docs.python.org/3/tutorial/datastructures.html#list-comprehensions)
+#[violation]
+pub struct UnpackedListComprehension;
+
+impl Violation for UnpackedListComprehension {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        "Replace unpacked list comprehension with a generator expression".to_string()
+    }
```

Comment:
I quite like the way we use `unreachable!()` in `and-or-tenary`:

https://github.com/astral-sh/ruff/blob/d8538d8c9816a4d09af6089e54c5232bae5998fc/crates/ruff_linter/src/rules/pylint/rules/and_or_ternary.rs#L34-L42